math for all THCuda*Tensor types (WORK IN PROGRESS)

wickedfoo opened the pull request:
Hey @soumith,

I've done most of the work to get fill() and zero() to work for all tensor types except CudaHalfTensor.

TensorMath.lua needs fixing in order to make them work for the other types. If you can get it to generate code for fill() and zero(), then I can take over again.

There is some hackiness due to CudaHalfTensor / __half at present, but will address soon.

Not all of the math functions have been generalized either.

You should be able to patch this in and update TensorMath.lua and anything else to get fill() and zero() to work for CudaIntTensor etc. though.

soumith commented:
got the functions working, have to get methods working, will finish up tomorrow.

soumith commented:
@wickedfoo here's the cwrap fixed up: https://github.com/torch/cutorch/tree/cutorch-template-types
It's in https://github.com/torch/cutorch in the branch cutorch-template-types
Commit / diff: https://github.com/torch/cutorch/commit/d5845a13437ecda03340bbc327be61e405246fd8

Tested that existing float stuff works, and the the double operations fill and zero work
```bash
th -lcutorch -e "a = torch.CudaDoubleTensor(10):fill(0); print(a)"
th -lcutorch -e "cutorch.test()"
```

wickedfoo commented:
diff updated with @soumith 's changes, and fixed a few other things.

CudaHalfTensor code will theoretically work now for the functions I've converted so far too, on CC 5.3+ and older devices with __half2float/__float2half

Will do a lot of cleanup, and maybe convert a small number of other things

soumith commented on lib/THC/THCTensorMasked.cu:
```diff
@@ -33,7 +33,7 @@ void THCudaTensor_maskedFill(THCState* state,
              2, "sizes do not match");
 
-  if (!THCudaTensor_pointwiseApply2(state, tensor, mask, TensorMaskedFillOp(value))) {
+  if (!cutorch_pointwiseApply2(state, tensor, mask, TensorMaskedFillOp(value))) {
```
rename this from cutorch_ to THC_

soumith commented on lib/THC/THCTensorMathPairwise.cu:
```diff
@@ -4,93 +4,97 @@
+struct TensorAddConstantOp<half> {
+  TensorAddConstantOp(half v) : val(v) {}
+  __device__ __forceinline__ void operator()(half* out, half* in) {
+#if __CUDA_ARCH__ >= 530
```
good stuff!

wickedfoo commented:
ok diff has been updated, fixed copy functions (conversions to/from different GPU tensor types happens exclusively on the GPU now)

moved all knowledge of HalfTensor/half compilation into THCHalf.h instead of all this CUDA_VERSION/__CUDA_ARCH__ crap

converted a few more functions

I should stop working on this for fear of making this diff too big.

wickedfoo commented on generic/CStorage.c:
```diff
@@ -66,7 +68,7 @@ static int cutorch_Storage_(copy)(lua_State *L)
   else if( (src = luaT_toudata(L, 2, "torch.CudaDoubleStorage")) )
     THCStorage_(copyCudaDouble)(state, storage, src);
-#if CUDA_VERSION >= 7050
+#ifdef CUDA_HALF_TENSOR
```
THCHalf.h defines this, and is the arbiter of everything half

wickedfoo commented on lib/THC/generic/THCTensorCopy.cu:
```diff
@@ -4,251 +4,44 @@
-      THCudaTensor_free(state, buffer);                                 \
-      THLongStorage_free(size);                                         \
-    }                                                                   \
+#define IMPLEMENT_THC_CUDA_TENSOR_COPY(TYPEC, TYPECUDA)                 \
```
copy from, say, CudaTensorDouble to CudaTensorByte now happens exclusively on the GPU

wickedfoo commented on lib/THC/THCTensorCopy.cu:
```diff
@@ -7,5 +7,226 @@ inline int curGPU() {
+};
+
+// Copy for the same type to the same type
+template <typename TensorTypeDst, typename TensorTypeSrc>
```
Most everything else in this diff is pretty mechanical; I think it would work if it compiled (excepting the half math implementations, which I may have messed up).

This function contains the one piece of real functionality change, since I now handle GPU -> GPU copies between different types exclusively on the GPU using this copy function.

This copy function is based off of the old GPU -> GPU copy function, but adds a case where the type of tensor differs between src and dst.

If there is one function worth looking at in detail and comparing against the old copy function, it is this one.

wickedfoo commented on lib/THC/THCApply.cuh:
```diff
@@ -164,55 +171,60 @@ bool THCudaTensor_pointwiseApply1(THCState* state,
+        break;                                  \
+        default:                                \
+        HANDLE_CASE(TYPE, -1);                  \
+        break;                                  \
```
I deleted the HANDLE_CASE(TYPE, 3) specializations from THCApply.cuh in an attempt to cut down on compile time (should address some of @hughperkins issues)

wickedfoo commented on lib/THC/generic/THCTensorCopy.cu:
```diff
@@ -4,251 +4,44 @@
+  THCTensor_(copyCuda##TYPEC)(THCState *state,                          \
+                              THCTensor *self,                          \
+                              THCuda##TYPECUDA##Tensor *src) {          \
+    cutorch_copyTensor<THCTensor, THCuda##TYPECUDA##Tensor>(state, self, src); \
```
i should rename this from cutorch_ -> THC_ too

wickedfoo commented:
Updated with []/__index/__newindex in Lua land working for all types (including CudaHalfTensor)
removed `hostreal` since the C API signature should be consistent; it's the Lua thunk responsibility to handle conversion for that type.

wickedfoo commented:
and fixed cutorch_copyTensor -> THC_copyTensor

soumith commented on lib/THC/THCTensorCopy.cu:
```diff
@@ -7,5 +7,226 @@ inline int curGPU() {
+  bool srcContig = TensorUtils<TensorTypeSrc>::isContiguous(state, src);
+  bool dstContig = TensorUtils<TensorTypeDst>::isContiguous(state, dst);
+  bool memcpyEligible =
+    ((srcContig && dstContig) || (totalElements == 1)) && sameType;
```
if totalElements is 1, then the tensor is contiguous anyways.

soumith commented:
sweet! merged.
